docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed cluster topologies.
+type ClusterClass struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+  // ControlPlane is a reference to a local struct that holds the details
+  // for provisioning the Control Plane for the Cluster.
+  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+  // Worker describes the worker nodes for the cluster.
+  // It is a collection of node types which can be used to create
+  // the worker nodes of the cluster.
+  // +optional
+  Worker WorkerClass `json:"worker,omitempty"`
+
+  // Infrastructure is a reference to a provider-specific template that holds
+  // the details for provisioning infrastructure specific cluster
+  // for the underlying provider.
+  // The underlying provider is responsible for the implementation
+  // of the template to an infrastructure cluster.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes which pronounce
+//
+type WorkerClass struct {
+  // Deployments is a list of machine deployment types that can be used to create
+  // a set of worker plane nodes.
+  Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+  // Class denotes a type of worker node present in the cluster
+  Class string `json:"class,omitempty"`
+
+
+  // Template is a local struct containing a collection of templates for creation of
+  // MachineDeployment objects representing a set of worker nodes.
+  Template TemplateClass `json:"template"`
+}
+
+type TemplateClass struct {
+  Metadata ObjectMeta `json:"metadata,omitempty"`
+
+  // Bootstrap contains the bootstrap template references to be used
+  // for the creation of worker Machines.
+  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+  // Infrastructure contains the infrastructure template references to be used
+  // for the creation of worker Machines.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+type LocalObjectTemplate struct {
+  // Ref is a required reference to a custom resource
+  // offered by a bootstrap provider.
+  Ref corev1.ObjectReference `json:"ref"`
+}
+```
+
+#### Modification to existing API Types
+##### Cluster
+1.  Add `Cluster.Spec.Managed` defined as
+    ```golang
+    // The information of the managed resources for the cluster.
+    // +optional
+    Managed *ManagedCluster `json:"managed,omitempty"`
```

Comment:
```suggestion
    Topology * ManagedTopology `json:"topology,omitempty"`
```

Reply:
I think the name of the field should still be `Managed` even if we update the name of the struct to `ManagedTopology`. I agree that the name **Managed Cluster** tends to be misleading, and **Managed Topology** seems to be an appropriate replacement. 

But the field named `Topology` does not convey that it is indeed managed by the Cluster object. Keeping the field name `Managed` does emphasize the managed aspect of the topology. wdyt?

Reply:
I missed this comment as it was hidden by an outdated / resolved block in the UI. I've added another comment in a similar theme: https://github.com/kubernetes-sigs/cluster-api/pull/4678#pullrequestreview-677406313.

If there are not other types of topologies, then it seems odd to draw this distinction.